cpp/src/lists/contains.cu

```diff
@@ -45,35 +46,81 @@ auto get_search_keys_device_iterable_view(cudf::scalar const& search_key, rmm::c
   return &search_key;
 }
 
-template <typename ElementType, bool has_nulls>
+// Extracts the ElementType::rep from decimals,
+// (e.g. to convert pair<decimal32, bool> => pair<int32_t, bool>).
+template <typename ElementType>
+struct transform_to_rep {
+  auto __device__ operator()(thrust::pair<ElementType, bool> element_and_validity) const
+  {
+    return thrust::make_pair(element_and_validity.first.value(), element_and_validity.second);
+  }
+};
+
+template <typename ElementType,
+          bool has_nulls,
+          std::enable_if_t<!cudf::is_fixed_point<ElementType>(), void>* = nullptr>
 auto get_pair_iterator(cudf::column_device_view const& d_search_keys)
 {
   return d_search_keys.pair_begin<ElementType, has_nulls>();
 }
 
-template <typename ElementType, bool>
+template <typename ElementType,
+          bool has_nulls,
+          std::enable_if_t<cudf::is_fixed_point<ElementType>(), void>* = nullptr>
+auto get_pair_iterator(cudf::column_device_view const& d_search_keys)
+{
+  return thrust::make_transform_iterator(d_search_keys.pair_begin<ElementType, has_nulls>(),
+                                         transform_to_rep<ElementType>{});
+}
+
+template <typename ElementType,
+          bool,
+          std::enable_if_t<!cudf::is_fixed_point<ElementType>(), void>* = nullptr>
 auto get_pair_iterator(cudf::scalar const& search_key)
 {
   return cudf::detail::make_pair_iterator<ElementType>(search_key);
 }
 
+template <typename ElementType,
+          bool,
+          std::enable_if_t<cudf::is_fixed_point<ElementType>(), void>* = nullptr>
+auto get_pair_iterator(cudf::scalar const& search_key)
```

Comment:
It seems like the `get_pair_iterator` shouldn't be necessary anymore and you can just use `make_pair_rep_iterator`?

Reply:
I've removed the `get_pair_iterator()` overloads for `scalar`, to use just one function.

We'd still need `get_pair_iterator()` to abstract how it's fetched for `column_view` input (i.e. `pair_begin()`) and scalar (i.e. `get_pair_rep_iterator()`)

Reply:
Does `make_pair_rep_iterator` not work on a `column_device_view` in addition to a `scalar`? Similar to how normal `make_pair_iterator` does?

Reply:
Not at the moment. Hmm... Is that something you'd recommend I pursue?

Reply:
Let me try get that going.

On a similar note, I currently have an iterator for iterating over elements in a list row (i.e. `list_device_view`). I'll look into making that iterable over `rep` values as well.

Reply:
I've eliminated the need for `get_pair_iterator()`. I need a moment to remove `get_list_iter()`, and retest.